Added Docker deal.II instructions to docs (#323)
lethe-cfd/lethe: doc/source/installation/docker.rst

Using a deal.II  Docker container
=================================

The deal.II library is a mature project containing `O(1,000,000)` lines of C++ code, itself depending on other similar libraries (e.g. Trilinos). Compiling deal.II with its exact dependencies can be a daunting task even for moderate C++ programmers - one with many knobs and dials, normally taking 4-6 hours.

Thankfully, the deal.II dev heroes also maintain some `Docker Hub <https://hub.docker.com/r/dealii/dealii/>`_ images containing pre-built deal.II libraries in a Ubuntu environment. For those unfamiliar with Docker, it allows the creation of completely isolated environments, most often pre-configured for specific apps; it is almost like a virtual machine, but running natively on your current OS (the technical term is *OS-level* virtualization, rather than *hardware-level*) - so scientific programs shouldn't have any degradation in performance. 

Below are some instructions for getting pre-configured deal.II environments, reducing the time needed to install lethe's dependencies from hours to minutes. There are 3 main parts in this tutorial:

> - If you have experience with Docker, the **0. TL;DR** section should be enough.

Reviewer: ```suggestion
- If you have experience with Docker, the :ref:`tl-dr` section should be enough.
```

- The next two sections have a brief introduction to Docker with enough information to get you started with deal.II containers.

> - The final **3. Saving Docker Commands in a Bash Script** section has a shell script you can run directly to start and manage deal.II containers.

Reviewer: ```suggestion
- The final :ref:`saving-docker` section has a shell script you can run directly to start and manage deal.II containers.
```

Sometimes it's easier to effectively set up a completely new OS than compile millions of lines of C++...

> 
> 
> 0. TL;DR
> --------

Reviewer: ```suggestion

.. _tl-dr:

0. TL;DR
--------
```

Reviewer: You can actually refer to subsections directly with hyperlinks in Sphinx, this is slightly cleaner than just using bold ;).

For example, on your local machine:

.. code-block:: bash

    local:~$ mkdir ~/lethe-simulations
    local:~$ cd ~/lethe-simulations
    local:~/lethe-simulations$ ls

    local$ docker run -it --name lethe-container -v $(pwd):/home/docker-host dealii/dealii:master-focal
    To run a command as administrator (user "root"), use "sudo <command>".
    See "man sudo_root" for details.

    dealii@2e92a38a6223:~$ cd /home/docker-host/
    dealii@2e92a38a6223:/home/docker-host$ echo "Hello lethe!" > somefile.txt
    dealii@2e92a38a6223:/home/docker-host$ exit

    local:~/lethe-simulations$ ls
    somefile.txt

That's all the Docker-specific tutorial! Launch your container running a deal.II image, go to :code:`/home/docker-host` to save your changes locally too, download lethe, compile it, and run your simulations there.

> 
> 3. Saving Docker Commands in a Bash Script
> ------------------------------------------

Reviewer: ```suggestion

.. _saving-docker:
3. Saving Docker Commands in a Bash Script
---------------------------------------------------------------
```
